[PyTorch] Devirtualize TensorImpl::dim() with macro

swolchok opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #50176 [PyTorch] Devirtualize TensorImpl::sizes() with macro
* **#49770 [PyTorch] Devirtualize TensorImpl::dim() with macro**
* #49767 [PyTorch] validate that SparseTensorImpl::dim needn't be overridden
* #49766 [PyTorch] Devirtualize TensorImpl::numel() with macro

Seems like the performance cost of making this commonly-called method virtual isn't worth having use of undefined tensors crash a bit earlier (they'll still fail to dispatch).

Differential Revision: [D25687465](https://our.internmc.facebook.com/intern/diff/D25687465/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D25687465/)!

swolchok committed:
[PyTorch] Devirtualize TensorImpl::dim() with macro

Seems like the performance cost of making this commonly-called method virtual isn't worth having use of undefined tensors crash a bit earlier (they'll still fail to dispatch).

Differential Revision: [D25687465](https://our.internmc.facebook.com/intern/diff/D25687465/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D25687465/)!

[ghstack-poisoned]

swolchok committed:
Update on "[PyTorch] Devirtualize TensorImpl::dim() with macro"


Seems like the performance cost of making this commonly-called method virtual isn't worth having use of undefined tensors crash a bit earlier (they'll still fail to dispatch).

Differential Revision: [D25687465](https://our.internmc.facebook.com/intern/diff/D25687465/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D25687465/)!

[ghstack-poisoned]

ezyang reviewed (approved):
OK. This change is a little hard to understand from the perspective of an OSS only user, since there is no point in the codebase where we ever actually set the macro. A Note explaining the situation would go a long way to explaining to people what's going on. (It's also worth pointing out XLA is the blocker for unconditionally removing the macro entirely)

swolchok committed:
Update on "[PyTorch] Devirtualize TensorImpl::dim() with macro"


Seems like the performance cost of making this commonly-called method virtual isn't worth having use of undefined tensors crash a bit earlier (they'll still fail to dispatch).

Differential Revision: [D25687465](https://our.internmc.facebook.com/intern/diff/D25687465/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D25687465/)!

[ghstack-poisoned]

swolchok commented:
>  A Note explaining the situation

Added to the PR at the bottom of the stack that adds `TENSORIMPL_MAYBE_VIRTUAL`.

swolchok committed:
Update on "[PyTorch] Devirtualize TensorImpl::dim() with macro"


Seems like the performance cost of making this commonly-called method virtual isn't worth having use of undefined tensors crash a bit earlier (they'll still fail to dispatch).

Differential Revision: [D25687465](https://our.internmc.facebook.com/intern/diff/D25687465/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D25687465/)!

[ghstack-poisoned]

swolchok committed:
Update on "[PyTorch] Devirtualize TensorImpl::dim() with macro"


Seems like the performance cost of making this commonly-called method virtual isn't worth having use of undefined tensors crash a bit earlier (they'll still fail to dispatch).

Differential Revision: [D25687465](https://our.internmc.facebook.com/intern/diff/D25687465/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D25687465/)!

[ghstack-poisoned]
